lib/pp.ml

```diff
@@ -297,3 +311,66 @@ let prvect_with_sep sep elem v = prvecti_with_sep sep (fun _ -> elem) v
 let prvect elem v = prvect_with_sep mt elem v
 
 let surround p = hov 1 (str"(" ++ p ++ str")")
+
+(*** DEBUG code ***)
+
+let db_print_pp ?(fmt=Format.std_formatter) pp =
+  let open Format in
+  let block_type btype =
+    let (bt, v) =
+      match btype with
+      | Pp_hbox v -> ("Pp_hbox", v)
+      | Pp_vbox v -> ("Pp_vbox", v)
+      | Pp_hvbox v -> ("Pp_hvbox", v)
+      | Pp_hovbox v -> ("Pp_hovbox", v)
+    in
+    Printf.sprintf "%s %d" bt v
+  in
+  let rec db_string_of_pp_r pp indent =
+    let ind () = fprintf fmt "%s" (String.make (2 * indent) ' ') in
+    ind();
```

Comment:
This is a wrong use of `Format`, please use the proper boxing and spacing operations.

Reply:
I can't get Format to indent correctly using boxes--I've been trying different things for a couple hours now.  Seems to me it's buggy and not well documented.  Also not getting the asprintf/sprintf to compile.

Any ideas?

```
let db_print_pp ?(fmt=Format.std_formatter) pp =
  let open Format in
  let block_type fmt btype =
    let (bt, v) =
      match btype with
      | Pp_hbox v -> ("Pp_hbox", v)
      | Pp_vbox v -> ("Pp_vbox", v)
      | Pp_hvbox v -> ("Pp_hvbox", v)
      | Pp_hovbox v -> ("Pp_hovbox", v)
    in
    fprintf fmt "%s %d" bt v
  in
  let rec db_print_pp_r pp =
    match pp with
    | Ppcmd_empty ->
        fprintf fmt "Ppcmd_empty@;"
    | Ppcmd_string str ->
        fprintf fmt "Ppcmd_string '%s'@;" str
    | Ppcmd_glue list ->
        fprintf fmt "Ppcmd_glue (@;";
        pp_open_vbox fmt 2;
        (*pp_print_break fmt 0 0;*)
        List.iter (fun x -> db_print_pp_r (repr x)) list;
        pp_close_box fmt ();
        fprintf fmt ") Ppcmd_glue@;"  (* todo: remove *)
    | Ppcmd_box (block, pp) ->
        fprintf fmt "Ppcmd_box %a@;" block_type block;
        db_print_pp_r (repr pp);
        pp_open_vbox fmt 2;
        pp_close_box fmt ()
    | Ppcmd_tag (tag, pp) ->
        fprintf fmt "Ppcmd_tag %s (@;" tag;
        (*pp_print_break fmt 0 0;*)
        pp_open_vbox fmt 2;
        db_print_pp_r (repr pp);
        pp_close_box fmt ();
        fprintf fmt ") Ppcmd_tag@;" (* TODO: remove *)
    | Ppcmd_print_break (i, j) ->
        fprintf fmt "Ppcmd_print_break %d %d@;" i j
    | Ppcmd_force_newline ->
        fprintf fmt "Ppcmd_force_newline@;"
    | Ppcmd_comment list ->
        fprintf fmt "Ppcmd_comment(@;";
        pp_open_vbox fmt 2;
        List.iter (fprintf fmt "%s@;") list;
        pp_close_box fmt ()
  in
  pp_open_vbox fmt 0;
  db_print_pp_r pp;
  pp_close_box fmt ();
  pp_print_flush fmt ()

let db_string_of_pp pp =
  (* Format.sprintf "%a" db_print_pp pp; todo: won't compile *)
  let buf = Buffer.create 256 in
  db_print_pp ~fmt:(Format.formatter_of_buffer buf) pp;
  Buffer.contents buf
```

Original and new output:

```
Ppcmd_glue (
  Ppcmd_glue (
    Ppcmd_tag start.diff.added.bg (
      Ppcmd_string ''
    ) Ppcmd_tag
    Ppcmd_string 'a'
    Ppcmd_string ' '
    Ppcmd_tag start.diff.added (
      Ppcmd_string 'b '
    ) Ppcmd_tag
    Ppcmd_tag end.diff.added (
      Ppcmd_string ' c'
    ) Ppcmd_tag
    Ppcmd_string ' '
    Ppcmd_string 'd'
    Ppcmd_string ' '
    Ppcmd_tag diff.added (
      Ppcmd_string 'e'
    ) Ppcmd_tag
    Ppcmd_string ' '
  ) Ppcmd_glue
  Ppcmd_tag end.diff.added.bg (
    Ppcmd_string ''
  ) Ppcmd_tag
) Ppcmd_glue

Ppcmd_glue (
Ppcmd_glue (
  Ppcmd_tag start.diff.added.bg (
    Ppcmd_string ''
      ) Ppcmd_tag
    Ppcmd_string 'a'
    Ppcmd_string ' '
    Ppcmd_tag start.diff.added (
    Ppcmd_string 'b '
      ) Ppcmd_tag
    Ppcmd_tag end.diff.added (
    Ppcmd_string ' c'
      ) Ppcmd_tag
    Ppcmd_string ' '
    Ppcmd_string 'd'
    Ppcmd_string ' '
    Ppcmd_tag diff.added (
    Ppcmd_string 'e'
      ) Ppcmd_tag
    Ppcmd_string ' '
    ) Ppcmd_glue
  Ppcmd_tag end.diff.added.bg (
  Ppcmd_string ''
    ) Ppcmd_tag
  ) Ppcmd_glue
```